### ISSUE TITLE
Make sure effects and other color management is being calculated in linear RGB space.

### ISSUE THREAD

**Issue by Giwayume** [viliusle/miniPaint#353]
Here's a fun fact, I found out that most CSS filters are hacks that don't do proper color space conversions that regular image editors do.

For example, css `filter: blur(5px)`. You can see a dark line in the middle where green and red are merged.

![image](https://user-images.githubusercontent.com/4075314/216755542-e5acebd6-e35b-460a-9865-b2fd9a3b09e2.png)

Here is the same image with a gaussian filter in GIMP.

![image](https://user-images.githubusercontent.com/4075314/216755579-445ad0fe-199e-4d92-bac0-fedf9352888d.png)

The reason behind this is most images are encoded in the sRGB color space. Working directly on pixels without converting sRGB to linear sRGB first results in mistakes like this. A typical editor workflow is:

1. Convert your source images from sRGB to linear sRGB
2. Run your edits/compositing operations using the linear sRGB colors
3. At the very end, convert linear sRGB back to sRGB for display.

I know the `hue-rotate` filter also has an issue where it is approximating the rotation in sRGB space instead of properly converting to HSV/HSL first, rotating, then converting back to sRGB. It may be beneficial to implement "slow, but correct" versions of these filters.

**Comment by Giwayume** [comment]
Note that the alpha blending of brushes also displays a similar issue, though this may be related to pre-multiplied alpha.

GIMP:

![image](https://user-images.githubusercontent.com/4075314/216791079-94f1e53e-af95-4cb3-8807-f84bb84825bf.png)

Minipaint:

![image](https://user-images.githubusercontent.com/4075314/216791045-96f14103-1b1c-45e7-8ec3-bf1d8c0a82f8.png)

**Comment by viliusle** [comment]
I need to do more research on it, and yes, it looks wrong. I just tried to do blur using CSS on chrome - same issue
![chrome](https://user-images.githubusercontent.com/3708786/216792396-1d32f9f8-64af-4640-b0eb-e9ea5865d910.png)

**Comment by Giwayume** [comment]
I can't believe there isn't a simple composite operation for "add". I'm having difficulty figuring out how the writers of the spec intended for us to be able to consume the pre-multiplied output of the path drawing operations.

**Comment by Giwayume** [comment]
Switching to the "display-p3" color space for the canvas in Chrome alleviates the issue somewhat, but it's not completely correct. The effect is just diminished due to the wider range of colors.

![image](https://user-images.githubusercontent.com/4075314/216802261-55741cab-4d91-4192-8952-059aabaa68f4.png)

**Comment by Giwayume** [comment]
I can't think of a way for the canvas (in realtime) to get the layer alpha blending working in linear space. It can definitely be done with WebGL fairly easily, but that would be quite a significant rewrite.

Specific things like the gaussian blur filter can be fixed, though, since most filters aren't realtime anyways. Maybe provide a toggle in the dialog for gaussian blur for "linear blending", but that no longer applies it as a realtime fx.

**Comment by Giwayume** [comment]
Here is an example I was testing image compositing in a Canvas renderer vs a WebGL renderer. The WebGL renderer composites images in linear sRGB color space, then gamma corrects at the very end back to non-linear sRGB. This gives the correct result.

Canvas renderer:

![image](https://user-images.githubusercontent.com/4075314/216841388-31590b98-ab10-47aa-bf19-eca04ef04115.png)

WebGL renderer:

![image](https://user-images.githubusercontent.com/4075314/216841331-da4f8ed1-17cf-4e50-abaf-cc2623769ed9.png)

**Comment by Giwayume** [comment]
https://www.youtube.com/watch?v=LKnqECcg6Gw

**Comment by viliusle** [comment]
Sadly, as I understand, WebGL would use different API than Canvas "2d". So it means almost rewriting most of miniPaint code.

**Comment by Giwayume** [comment]
Yes, WebGL is much more complicated to set up. But it is also much more performant and gives the ability to do realtime effects with shaders. May be something to consider.
